### DISCUSSION TITLE
Execute 3 different jobs after the first one finished

### DISCUSSION THREAD

**Discussion by advapiIT** [quartznet/quartznet#1777] [Q&A]
Hello,
I've used in the past Quartz.net to schedule single jobs.. now I've been asked to perform a first step that downloads data from a site (climate data) then saves them inside a file. this works great with Quartz and no problem.

Then I've to execute other 3 jobs that would procede subsets of those data, each one with it's  logic and can be executed indipendently.

I'm under .NET 6 and using Quartz.Extensions.Hosting to use DI.

For developing purpose I've done really ugly things doing at the end of the main job

```c#
 context.MergedJobDataMap["data"] = JsonConvert.SerializeObject(jobPayload);

      ProcessFistTypeObjectsJob tmp = new ProcessFistTypeObjectsJob(_configuration, _logger, _historicalRunProcessorService, _metadataInformationService, _downloadFileService);

      await tmp.Execute(context);
```

I know I'm an ugly person.

My programs.cs is has

```c#
 var host = Host.CreateDefaultBuilder(args).ConfigureServices(services =>
        {
          // Add the required Quartz.NET services
          services.AddQuartz(q =>
          {
            q.UseMicrosoftDependencyInjectionJobFactory();

            // Register the job, loading the schedule from configuration
            q.AddJobAndTrigger<PerformOperationsJob>(configuration);
          });

          // Add the Quartz.NET hosted service

          services.AddQuartzHostedService(
            q => q.WaitForJobsToComplete = true);

```

Each job has 

```c#
 [DisallowConcurrentExecution]
  [PersistJobDataAfterExecution]
  public class PerformOperationsJob : IJob
  {

```
as implementation.

I've googled around and tried either the JobListener or this [link ](https://www.codeproject.com/Articles/1201922/Dependent-job-scheduling-with-Quartz-NET) but with no luck...

Anyone can help me?
Thanks

**Comment by lahma** [comment]
You should always add information that you have cross-posted same question:

* https://stackoverflow.com/questions/73945135/starting-2-different-jons-when-a-main-one-finished-in-quartz-net
* https://stackoverflow.com/questions/73950301/using-a-producer-consumer-pattern-with-quartz-net

**Comment by advapiIT** [comment]
Ops... I tought it and forgot it.. sorry..

**Reply by lahma** [reply]
If you are not going for a queue system like MassTransit, I would probably trigger the next steps using `scheduler.TriggerJob` and passing target as job data map paramater (file/data id ).

**Comment by advapiIT** [comment]
But I ve tried a similar thing but the main job is executed even if it's
doing the other jobs... and it's not possibile since woukd lead to errors

Il Mer 5 Ott 2022, 18:16 Marko Lahma ***@***.***> ha scritto:

**Reply by lahma** [reply]
All guesswork here, publish a full repro or otherwise it's pretty impossible to determine what could be the cause.

**Reply by advapiIT** [reply]
Hello @lahma , here it's https://github.com/advapiIT/QuartzTest

Just Run it... you'll see that the main job executes, then when the second starts, again the first re-execute due to chon expression elapsed.

Thanks

**Reply by lahma** [reply]
I think that code behaves as expected. You main job exists after 10seconds and after that it's free to be executed again based on schedule. The `DisallowConcurrentExecution` is only protecting the main job so anything triggered after it won't share the same lock. 

I think it would be easier to design system in a way that each download is separate trigger and passes name of the file along the chain. Then there can be multiple files in processing but chain would only handle single file and thus would be free of concurrency issues. Main job could poll the files and then trigger first "new main" with file name/id in trigger data map.

**Comment by advapiIT** [comment]
So you tell me to have some resident threads that receive data from a queue
and process them while the only main triggerable job (scheduled via chron)
is fired each x time.

Thanks

**Reply by lahma** [reply]
"Resident thread" can also be a polling job. But pushing trigger for each file seems reasonable if you want to go that way. You can play with different approaches.

**Comment by advapiIT** [comment]
you mean a thread that pool on a queue, with a spin-down

Thanks
Paolo

**Reply by lahma** [reply]
I think I'll have to let you do the design here 🙂
